gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/HttpHandlerUtil.java

```diff
@@ -71,9 +74,32 @@ public class HttpHandlerUtil {
     /**
      * Convert a http request into a {@link RequestMessage}.
      */
-    public static RequestMessage getRequestMessageFromHttpRequest(final FullHttpRequest request) {
-        final String contentType = Optional.ofNullable(request.headers().get(HttpHeaderNames.CONTENT_TYPE)).orElse("application/json");
+    public static RequestMessage getRequestMessageFromHttpRequest(final FullHttpRequest request,
+                                                                  Map<String, MessageSerializer<?>> serializers) throws SerializationException {
+        final String contentType = request.headers().get(HttpHeaderNames.CONTENT_TYPE);
+
+        if (!contentType.equals("application/json") && serializers.containsKey(contentType)) {
```

Comment:
the `!contentType.equals("application/json")` condition means that you use a naked JSON `ObjectMapper` to process the request. with the `Accept` header a user could specify "application/json" to mean Typed GraphSON 3 (for example) if the server was not configured that way:

https://github.com/apache/tinkerpop/blob/master/gremlin-util/src/main/java/org/apache/tinkerpop/gremlin/util/ser/GraphSONMessageSerializerV3.java#L85

I guess for the `Content-Type` and the HTTP specification it's expected that the client be explicit in terms of what the payload is and there is no form of content negotiation that goes on there.  I suppose that having it work this way spares the user from having to properly configure the server, such that `application/json` ties to a Untyped GraphSON serializer (which i can only presume would work as well as a naked `ObjectMapper`).  

could you test to see if that is the case? if the `GraphSONUntypedMessageSerializerV3` decodes the request in the same fashion that the naked `ObjectMapper` does, then I think it would remove doubts about any inconsistencies we'd need to warn users about. Moreover, we could technically remove the naked` ObjectMapper` and this condition and just try to decode with the serializers. 

Perhaps said another way, instead of using `!contentType.equals("application/json")` to control flow here, use the presence/absence of a header in the payload to do that, then if it is there  you just use the `MessageSerializer` and if it is not you pull the `Mapper` object from the serializer:

https://github.com/apache/tinkerpop/blob/master/gremlin-util/src/main/java/org/apache/tinkerpop/gremlin/util/ser/AbstractGraphSONMessageSerializerV2.java#L162-L165

and use that to decode the payload. I think one neat thing about this approach is that you could conceivable then send your payload with or without the header potentially (though the underlying `Mapper` object between GraphSON and GraphBinary don't have the same interfaces but theoretically it seems like it could work). 

one other point here perhaps it would be good to check the verb earlier to avoid this path: `request.method() == POST`?

Reply:
what types of requests I think to support:
1. existing https://tinkerpop.apache.org/docs/current/reference/#connecting-via-http
intended to use with curl, postman, etc by users
both GET and POST
`Content-Type` header can be empty or `application/json`
`Accept`  header can be any, but I think most useful can be `application/json`, `text/plain`, `application/vnd.gremlin-v3.0+json` and `application/vnd.gremlin-v3.0+json;types=false`
Request body example: `{ "gremlin": "g.V()" }`
We can introduce some model to deserialize with object mapper or continue to use [existing code](https://github.com/apache/tinkerpop/blob/35ac4963f72699dd123fbc99022824df37988f94/gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/HttpHandlerUtil.java#L74)

2. newly introduced payload with serialized `RequestMessage`
intended for drivers/GLV's. Support both gremlin and bytecode queries.
only POST
`Content-Type` is defined by used serializer, expected type GraphSON `application/vnd.gremlin-v3.0+json` or GraphBinary `application/vnd.graphbinary-v1.0`. Untyped GraphSON is not supported, it can't deserialize bytecode
`Accept`  header can be any. I may add validation that used same serializer for request and response.
Request body contains serialized `RequestMessage`

To distinguish this types of requests I propose to use [Content-Type HTTP header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type). I think that using a payload to determine the data type is an unnecessary complication and not HTTP way.

`GraphSONUntypedMessageSerializerV3` can't deserialize bytecode request, it knows nothing about steps or parameters, everything turns into HashMap of HashMap.
![image](https://github.com/apache/tinkerpop/assets/7605999/ec41eba1-89fe-4ef3-81f6-e6783783ca21)

Reply:
It seems strange to me to use the `Content-Type` for anything more than a serialization rule. If I understand correctly, by your definition of what you want to support, you can only access the old HTTP API with `application/json` but if you make it `application/vnd.gremlin-v3.0+json` then the payload must be a `RequestMessage` or else it won't work. I suppose we didn't support `application/vnd.gremlin-v3.0+json` as a `Content-Type` there before so maybe that's ok, but now that there is a path that allows you to POST that format, it wouldn't surprise me if folks tried it, especially since from the perspective `Accept` you can request `application/json` and assuming you've configured the server that way you could get back a `application/vnd.gremlin-v3.0+json` formatted payload. 

well, please think through the validation on requests here carefully and write good documentation to clarify how this works. we need to make sure folks understand the API they are calling and what `Content-Type` and `Accept` headers the need to use to get the results that they want.